### ISSUE TITLE
Opting to use another solution than foreign-generics?

### ISSUE THREAD

**Issue by justinwoo** [purescript-aws-sdk/gen#10]
The generics machinery has some cost in runtime and it looks like all the data types are newtyped records. Would you consider maybe simply exporting the record types directly?

This way, users could opt into using libraries they want like [Simple-JSON](https://github.com/justinwoo/purescript-simple-json/) and [Record-Show](https://pursuit.purescript.org/packages/purescript-record-show/0.4.0) and whatnot (though in the Show case, this would not work with nested records for now)

**Comment by plippe** [comment]
Hey,

100% agree that forcing any encoding / decoding library, like https://github.com/paf31/purescript-foreign-generic, is the wrong move. The long term wish is to have multiple projects, very generic core, and one implementation of encoding / decoding. This would allow users to create their own encoders / decoders if our version didn't please them.

I am not sure how to handle that at the moment. Decoupling the objects from their foreign reader / writer is doable, but the amount of projects would double (~130 to ~260). Not the end of the work, but definitely something unpleasant to manage.

About `newtype`, with only a few weeks of PureScript under my belt, I didn't see anything better to use. I wanted a single constructor, no overhead once converted to JS, and type safety. `data`, `type`, and `records` + primitive types didn't seem to be the right approach. 

I would be very happy to hear your opinions on the matter.

**Comment by justinwoo** [comment]
I think people who want them will define their own newtype in any case and use whatever library they wish. Currently, you're having to require foreign-generic and provide decoding instances for them, as you can't have orphan instances. But by exposing the type aliases directly, users can choose to do whatever they want, e.g. with Simple-JSON:

```hs
-- from your lib
type Fruit = { apple :: String }

-- Simple-JSON
readFruit :: Foreign -> Either _ Fruit
readFruit = readJSON -- concrete records can be automatically derived with simple-json

newtype FruitNT = FruitNT Fruit
derive newtype instance readForeignFruitNT :: ReadForeign FruitNT
```

For normal PureScript-level functions though, it would be better to expose functions that use newtypes and data types as necessary to make functions well-typed, e.g. newtype Id, data Choice, etc.

When you have all this normally, then your users can also either use the functions you expose for `Fruit` directly, or use Newtype instances (`derive instance newtypeFruitNT :: Newtype Fruit _`) to use the [over](https://pursuit.purescript.org/packages/purescript-newtype/2.0.0/docs/Data.Newtype#v:over) function for lifting your exposed functions to what they need.

**Comment by plippe** [comment]
Looks good, but I still have a question.

Currently the encoding / decoding is done in the [request](https://github.com/purescript-aws-sdk/purescript-aws-request/blob/master/src/AWS/Request.purs). I don't believe this would be doable with type classes anymore.

I see an easy alternative by passing encoder and decoder functions. 

```purescript
request :: forall eff a b c. 
  (a -> Foreign) -> 
  (Foreign -> Either c b) -> 
  a -> 
  Aff (exception :: EXCEPTION | eff) b

fruitRequest :: forall eff c. 
  (FruitRequest -> Foreign) -> 
  (Foreign -> Either c FruitResponse) ->
  FruitRequest ->
  Aff (exception :: EXCEPTION | eff) FruitResponse
```

Is that what you had in mind ?

**Comment by justinwoo** [comment]
Yeah, I think this makes the most sense, and then users can define their own wrapper functions for whatever libraries they're using.

**Comment by plippe** [comment]
Will check with the team. If there aren't any strong objections, will move in that direction. 
Do you mind if I add you to the relevant PRs ?

**Comment by justinwoo** [comment]
Np, feel free to link them on #purescript on the FP slack too (https://fpchat-invite.herokuapp.com/)

**Comment by justinwoo** [comment]
Could you explain why you would need extra projects and why these concretely typed (Foreign -> Either _ a) functions can't be included in existing projects?

**Comment by plippe** [comment]
At the moment, we are looking to make it easy for everyone to just start using this library. Removing foreign-generic would force users to bring their own de / encoding library, and for us to generate another set of projects with the de / encoders.

For comparison, the aws-sdk doesn't allow you to pick which library they use, we are just doing the same.

**Comment by plippe** [comment]
Dustin Whitney (I don't have a handle) made a strong point about using [simple-json](https://github.com/justinwoo/purescript-simple-json/) over (foreign-generic)[https://github.com/paf31/purescript-foreign-generic]. 

Will look into migrating over the next few weeks.
